File: desktop/packages/core/src/data/session-repository.ts
import type { Database } from '../db/connection.js';
import type { ActiveSession } from '../domain/active-session.js';
import { localUserId } from '../ids.js';
import { fromEpochSeconds, msToSeconds, toEpochSeconds } from '../time.js';
import type { HistoryEntry, SessionDetailView, SessionRepository } from './ports.js';

interface RawSurvey {
  focus_rating: number;
  comprehension_rating: number | null;
  difficulty_rating: number | null;
  note: string | null;
}

interface RawInterruption {
  kind: string;
  occurred_at: number;
  duration_s: number | null;
}

interface RawHistory {
  id: string;
  subject_name: string;
  started_at: number;
  actual_duration_s: number | null;
  end_reason: string | null;
}

interface RawOpenSession {
  id: string;
  started_at: number;
  updated_at: number;
  paused_duration_s: number;
}

export class SqliteSessionRepository implements SessionRepository {
  constructor(
    private readonly db: Database,
    private readonly now: () => Date = () => new Date(),
  ) {}

  async insertStarted(s: ActiveSession): Promise<void> {
    const ts = toEpochSeconds(this.now());
    this.db
      .prepare(
        `INSERT INTO sessions
           (id, user_id, subject_id, mode, started_at, created_at, updated_at)
         VALUES (?, ?, ?, ?, ?, ?, ?)`,
      )
      .run(s.id, localUserId, s.subjectId, s.mode, toEpochSeconds(s.startedAt), ts, ts);
    return;
  }

  /**
   * Written on pause AND resume: every state change is persisted
   * (architecture.md §3.4). On pause the accumulated value is unchanged, but
   * `updated_at` advances — and that watermark is what makes crash recovery's
   * last-write bound exact for a session that dies while paused.
   *
   * `AND ended_at IS NULL` is the guard: an ended session is immutable at the
   * DATA layer, not merely by caller discipline (V3 spec §4.3).
   */
  async updatePausedDuration(id: string, totalPausedMs: number): Promise<void> {
    this.db
      .prepare(
        `UPDATE sessions
            SET paused_duration_s = ?, updated_at = ?
          WHERE id = ? AND ended_at IS NULL`,
      )
      .run(msToSeconds(totalPausedMs), toEpochSeconds(this.now()), id);
    return;
  }

  /** The one-shot closing write, guarded the same way. */
  async end(a: {
    id: string;
    endedAt: Date;
    actualDurationMs: number;
    totalPausedMs: number;
  }): Promise<void> {
    this.db
      .prepare(
        `UPDATE sessions
            SET ended_at = ?, actual_duration_s = ?, paused_duration_s = ?,
                end_reason = 'user_ended', updated_at = ?
          WHERE id = ? AND ended_at IS NULL`,
      )
      .run(
        toEpochSeconds(a.endedAt),
        msToSeconds(a.actualDurationMs),
        msToSeconds(a.totalPausedMs),
        toEpochSeconds(this.now()),
        a.id,
      );
    return;
  }

  /**
   * Closes out sessions left open by a crash (architecture.md §3.4).
   *
   * `ended_at` is the row's last persisted write — an honest lower bound, not a
   * guess — and `end_reason = 'crashed'` is the key streak inputs exclude on,
   * because a crashed session's duration cannot be trusted to weight a
   * qualified day (data-model.md §5.1).
   *
   * LAUNCH-TIME ONLY. "Unterminated" is the only signal schema v1 carries, so a
   * *running* session is indistinguishable from a crashed one and would be
   * closed too. Call this once, before any session can start — never on a timer.
   */
  async recoverCrashedSessions(): Promise<number> {
    const count = this.db.transaction(() => {
      const open = this.db
        .prepare(
          `SELECT id, started_at, updated_at, paused_duration_s
             FROM sessions
            WHERE ended_at IS NULL`,
        )
        .all<RawOpenSession>();
      const nowS = toEpochSeconds(this.now());
      for (const r of open) {
        const activeS = r.updated_at - r.started_at - r.paused_duration_s;
        // The ended_at guard is redundant inside this transaction; it keeps the
        // write safe if this loop is ever run outside one.
        this.db
          .prepare(
            `UPDATE sessions
                SET ended_at = ?, actual_duration_s = ?, end_reason = 'crashed', updated_at = ?
              WHERE id = ? AND ended_at IS NULL`,
          )
          .run(r.updated_at, Math.max(0, activeS), nowS, r.id);
      }
      return open.length;
    });
    return count;
  }

  /**
   * Note the join does NOT filter subjects on `deleted_at`: a soft-deleted or
   * archived subject must still name its sessions, which is the entire reason
   * subject deletion is soft.
   */
  async listHistory(): Promise<HistoryEntry[]> {
    const rows = this.db
      .prepare(
        `SELECT s.id, sub.name AS subject_name, s.started_at,
                s.actual_duration_s, s.end_reason
           FROM sessions s
           JOIN subjects sub ON sub.id = s.subject_id
          WHERE s.deleted_at IS NULL AND s.ended_at IS NOT NULL
          ORDER BY s.started_at DESC, s.id DESC`,
      )
      .all<RawHistory>();
    return rows.map((r) => ({
      sessionId: r.id,
      subjectName: r.subject_name,
      startedAt: fromEpochSeconds(r.started_at),
      actualDurationS: r.actual_duration_s ?? 0,
      endReason: r.end_reason,
    }));
  }

  /**
   * The survey (if the session was rated) plus its event log.
   *
   * This is the ONE legal reader of the event log outside its own repository,
   * and `test/db/write-confinement.test.ts` names this file explicitly — a
   * second reader has to be added there deliberately, in front of a reviewer.
   */
  async loadSessionDetail(sessionId: string): Promise<SessionDetailView> {
    const survey = this.db
      .prepare(
        `SELECT focus_rating, comprehension_rating, difficulty_rating, note
           FROM session_surveys
          WHERE session_id = ? AND deleted_at IS NULL`,
      )
      .get<RawSurvey>(sessionId);

    // Ordered by id as well as time: occurred_at is epoch SECONDS, so two
    // events in the same second would otherwise sort arbitrarily between runs.
    // newId() is UUIDv7, which is lexicographically time-ordered.
    const rows = this.db
      .prepare(
        `SELECT kind, occurred_at, duration_s
           FROM interruptions
          WHERE session_id = ? AND deleted_at IS NULL
          ORDER BY occurred_at ASC, id ASC`,
      )
      .all<RawInterruption>(sessionId);

    return {
      interruptions: rows.map((r) => ({
        kind: r.kind,
        occurredAt: fromEpochSeconds(r.occurred_at),
        durationS: r.duration_s,
      })),
      focusRating: survey?.focus_rating ?? null,
      comprehensionRating: survey?.comprehension_rating ?? null,
      difficultyRating: survey?.difficulty_rating ?? null,
      note: survey?.note ?? null,
      hasSurvey: survey !== undefined,
    };
  }
}

File: app-flutter/lib/features/session/data/session_repository.dart
import 'package:drift/drift.dart';

import '../../../core/db/database.dart';
import '../../../core/ids.dart';
import '../domain/active_session.dart';
import 'interruption_repository.dart';

class HistoryEntry {
  const HistoryEntry({
    required this.sessionId,
    required this.subjectName,
    required this.startedAt,
    required this.actualDuration,
    this.endReason,
  });

  final String sessionId;
  final String subjectName;
  final DateTime startedAt;
  final Duration actualDuration;
  final String? endReason;
}

/// Everything the Stats inline expansion shows for one session: its survey (if
/// it was rated) and its interruption log.
class SessionDetailView {
  const SessionDetailView({
    required this.interruptions,
    this.focusRating,
    this.comprehensionRating,
    this.difficultyRating,
    this.note,
  });

  final List<InterruptionEntry> interruptions;
  final int? focusRating;
  final int? comprehensionRating;
  final int? difficultyRating;
  final String? note;

  bool get hasSurvey => focusRating != null;
}

class SessionRepository {
  SessionRepository(this._db);

  final AppDatabase _db;

  Future<SessionDetailView> loadSessionDetail(String sessionId) async {
    final survey = await (_db.select(_db.sessionSurveys)
          ..where((t) => t.sessionId.equals(sessionId) & t.deletedAt.isNull()))
        .getSingleOrNull();
    // Ordered by id as well as time: drift stores DateTime as epoch SECONDS, so
    // two events in the same second would otherwise sort arbitrarily between
    // runs. newId() is UUIDv7, which is lexicographically time-ordered.
    final rows = await (_db.select(_db.interruptions)
          ..where((t) => t.sessionId.equals(sessionId) & t.deletedAt.isNull())
          ..orderBy([
            (t) => OrderingTerm.asc(t.occurredAt),
            (t) => OrderingTerm.asc(t.id),
          ]))
        .get();
    return SessionDetailView(
      interruptions: [
        for (final r in rows)
          InterruptionEntry(
              kind: r.kind, occurredAt: r.occurredAt, durationS: r.durationS),
      ],
      focusRating: survey?.focusRating,
      comprehensionRating: survey?.comprehensionRating,
      difficultyRating: survey?.difficultyRating,
      note: survey?.note,
    );
  }

  Future<void> insertStartedSession(ActiveSession s) async {
    await _db.into(_db.sessions).insert(SessionsCompanion.insert(
          id: s.id,
          userId: localUserId,
          subjectId: s.subjectId,
          mode: s.mode,
          startedAt: s.startedAt,
        ));
  }

  Future<void> updatePausedDuration(String id, Duration totalPaused) async {
    // Guarded: an ended session is immutable (data-model.md §3.4).
    await (_db.update(_db.sessions)
          ..where((t) => t.id.equals(id) & t.endedAt.isNull()))
        .write(
      SessionsCompanion(
        pausedDurationS: Value(totalPaused.inSeconds),
        updatedAt: Value(DateTime.now().toUtc()),
      ),
    );
  }

  Future<void> endSession({
    required String id,
    required DateTime endedAt,
    required Duration actualDuration,
    required Duration totalPaused,
  }) async {
    // Guarded: ending is a one-shot closing write (data-model.md §3.4).
    await (_db.update(_db.sessions)
          ..where((t) => t.id.equals(id) & t.endedAt.isNull()))
        .write(
      SessionsCompanion(
        endedAt: Value(endedAt),
        actualDurationS: Value(actualDuration.inSeconds),
        pausedDurationS: Value(totalPaused.inSeconds),
        endReason: const Value('user_ended'),
        updatedAt: Value(DateTime.now().toUtc()),
      ),
    );
  }

  /// Closes out sessions left open by a crash (architecture.md §3.4).
  /// ended_at is the row's last persisted write — an honest lower bound —
  /// and end_reason 'crashed' is the key streak inputs will exclude on.
  /// Returns the number of sessions recovered.
  Future<int> recoverCrashedSessions() async {
    return _db.transaction(() async {
      final open = await (_db.select(_db.sessions)
            ..where((t) => t.endedAt.isNull()))
          .get();
      for (final row in open) {
        final active = row.updatedAt.difference(row.startedAt).inSeconds -
            row.pausedDurationS;
        // The isNull guard is redundant inside this transaction; it keeps the
        // write safe if this loop ever runs outside one.
        await (_db.update(_db.sessions)
              ..where((t) => t.id.equals(row.id) & t.endedAt.isNull()))
            .write(
          SessionsCompanion(
            endedAt: Value(row.updatedAt),
            actualDurationS: Value(active < 0 ? 0 : active),
            endReason: const Value('crashed'),
            updatedAt: Value(DateTime.now().toUtc()),
          ),
        );
      }
      return open.length;
    });
  }

  Stream<List<HistoryEntry>> watchHistory() {
    final query = _db.select(_db.sessions).join([
      innerJoin(
          _db.subjects, _db.subjects.id.equalsExp(_db.sessions.subjectId)),
    ])
      ..where(
          _db.sessions.deletedAt.isNull() & _db.sessions.endedAt.isNotNull())
      ..orderBy([OrderingTerm.desc(_db.sessions.startedAt)]);
    return query.watch().map((rows) => rows.map((row) {
          final session = row.readTable(_db.sessions);
          final subject = row.readTable(_db.subjects);
          return HistoryEntry(
            sessionId: session.id,
            subjectName: subject.name,
            startedAt: session.startedAt,
            actualDuration: Duration(seconds: session.actualDurationS ?? 0),
            endReason: session.endReason,
          );
        }).toList());
  }
}
